src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1852,10 +1878,113 @@ ucs_status_t uct_ib_mlx5_devx_mem_attach(uct_md_h uct_md,
     return status;
 }
 
+static ucs_status_t
+uct_ib_mlx5_devx_device_mem_alloc(uct_md_h uct_md, size_t *length_p,
+                                  void **address_p, ucs_memory_type_t mem_type,
+                                  unsigned flags, const char *alloc_name,
+                                  uct_mem_h *memh_p)
+{
+#if HAVE_IBV_DM && HAVE_DECL_IBV_IMPORT_DM
+    uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
+    uct_md_mem_reg_params_t reg_params = {
+        .field_mask = 0
+    };
+
+    /* Align the allocation to a potential use of registration cache */
+    size_t length = ucs_align_up_pow2(*length_p, md->alloc_align);
+    struct ibv_alloc_dm_attr dm_attr;
+    struct ibv_dm *dm;
+    uct_ib_mlx5_devx_mem_t *memh;
+    void *address;
+    ucs_status_t status;
+    unsigned mem_flags = UCT_MD_MEM_ACCESS_REMOTE_GET |
+                         UCT_MD_MEM_ACCESS_REMOTE_PUT |
+                         UCT_MD_MEM_ACCESS_REMOTE_ATOMIC;
```

Comment:
Maybe do initializations lazy (after `if` blocks which might lead to exit before these values are used at all)